usr.bin/sockstat/sockstat.c

```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+			if (laddr->address.ss_len > 0)
+				formataddr(&laddr->address, buf, bufsize);
+			else
+				strlcpy(buf, "??", bufsize);
+			printf(" %-*s", cw->local_addr, buf);
+			format_unix_faddr(faddr, buf, bufsize);
+			printf(" %-*s", cw->foreign_addr, buf);
+		} else {
+			if (laddr != NULL) 
+				formataddr(&laddr->address, buf, bufsize);
+			else
+				strlcpy(buf, "??", bufsize);
+			printf(" %-*s", cw->local_addr, buf);
+			if (faddr != NULL)
+				formataddr(&faddr->address, buf, bufsize);
+			else
+				strlcpy(buf, "??", bufsize);
+			printf(" %-*s", cw->foreign_addr, buf);
+		}
+		if (opt_A)
+			printf(" 0x%16lx", s->pcb);
+		if (opt_f) 
+			printf(" %*d", cw->fib, s->fibnum);
+		if (opt_I) {
+			if (s->splice_socket != 0) {
+				struct sock *sp;
+				sp = RB_FIND(socks_t, &socks, &(struct sock)
+					{ .socket = s->splice_socket });
+				if (sp != NULL)
+					formataddr(&sp->laddr->address,
+								buf, bufsize);
+			}
+			else
+				strlcpy(buf, "??", bufsize);
+			printf(" %-*s", cw->splice_address, buf);
+		}
+		if (opt_i) {
+			if (s->proto == IPPROTO_TCP ||
+				s->proto == IPPROTO_UDP) {
+				printf(" %*" PRIu64, cw->inp_gencnt ,s->inp_gencnt);
+			}
+			else
+				printf(" %*s", cw->inp_gencnt, "??");
+		}
+		if (opt_U) {
+			if (faddr != NULL &&
+				((s->proto == IPPROTO_SCTP &&
+					s->state != SCTP_CLOSED &&
+					s->state != SCTP_BOUND &&
+					s->state != SCTP_LISTEN) ||
+					(s->proto == IPPROTO_TCP &&
+					s->state != TCPS_CLOSED &&
+					s->state != TCPS_LISTEN))) {
+				printf(" %*u", cw->encaps, ntohs(faddr->encaps_port));
+			}
+			else
+				printf(" %*s", cw->encaps, "??");
+		}
+		if (opt_s) {
+			if (faddr != NULL &&
+				s->proto == IPPROTO_SCTP &&
+				s->state != SCTP_CLOSED &&
+				s->state != SCTP_BOUND &&
+				s->state != SCTP_LISTEN) {
+				printf(" %-*s", cw->path_state,
+					sctp_path_state(faddr->state));
+			}
+			else
+				printf(" %-*s", cw->path_state, "??");
+		}
+		if (first) {
+			if (opt_s) {
+				if (s->proto == IPPROTO_SCTP ||
+					s->proto == IPPROTO_TCP) {
+					switch (s->proto) {
+					case IPPROTO_SCTP:
+						printf(" %-*s", cw->conn_state,
+							sctp_conn_state(s->state));
+						break;
+					case IPPROTO_TCP:
+						if (s->state >= 0 &&
+							s->state < TCP_NSTATES)
+							printf(" %-*s", cw->conn_state,
+								tcpstates[s->state]);
 						else
-							pos += xprintf("?");
+							printf(" %-*s", cw->conn_state, "??");
 						break;
 					}
 				}
-				offset += 13;
+				else
+					printf(" %-*s", cw->conn_state, "??");
 			}
 			if (opt_S) {
-				if (s->proto == IPPROTO_TCP) {
-					do
-						pos += xprintf(" ");
-					while (pos < offset);
-					pos += xprintf("%.*s",
-					    TCP_FUNCTION_NAME_LEN_MAX,
-					    s->stack);
-				}
-				offset += TCP_FUNCTION_NAME_LEN_MAX + 1;
+				if (s->proto == IPPROTO_TCP)
+					printf(" %-*.*s", TCP_FUNCTION_NAME_LEN_MAX,
+						TCP_FUNCTION_NAME_LEN_MAX, s->stack);
+				else
+					printf(" %-*s", cw->stack, "??");
```

Comment:
For both brevity and consistency
```suggestion
					printf(" %-*.*s", cw->stack, cw->stack, s->stack);
				else
					printf(" %-*s", cw->stack, "??");
```